staging/src/k8s.io/apimachinery/pkg/api/meta/help.go

```diff
@@ -127,7 +127,16 @@ func EachListItemWithAlloc(obj runtime.Object, fn func(runtime.Object) error) er
 // allocNew: Whether shallow copy is required when the elements in Object.Items are struct
 func eachListItem(obj runtime.Object, fn func(runtime.Object) error, allocNew bool) error {
 	if unstructured, ok := obj.(runtime.Unstructured); ok {
-		return unstructured.EachListItem(fn)
+		final := fn
+		if allocNew {
+			final = func(obj runtime.Object) error {
```

Comment:
hmm... in the case of Unstructured#EachListItem, we're actually already given a standalone copy of the item, so this change is *increasing* cost. How much more does using reflection cost us here?

I'm wondering about adding `EachListItemWithAlloc(func(Object) error) error` to `type Unstructured interface`

Reply:
```bash
benchcmp /tmp/meta_helper.old /tmp/meta_helper.new

benchmark                                              old ns/op     new ns/op     delta
BenchmarkEachListItem/FooList-16                       34788         35147         +1.03%
BenchmarkEachListItem/FooList-16                       35425         34933         -1.39%
BenchmarkEachListItem/SampleList-16                    39058         39419         +0.92%
BenchmarkEachListItem/SampleList-16                    39355         39551         +0.50%
BenchmarkEachListItem/RawExtensionList-16              24741         24747         +0.02%
BenchmarkEachListItem/RawExtensionList-16              24747         24913         +0.67%
BenchmarkEachListItem/UnstructuredList-16              3076          2619          -14.86%
BenchmarkEachListItem/UnstructuredList-16              2883          2468          -14.39%
BenchmarkEachListItemWithAlloc/FooList-16              187323        187752        +0.23%
BenchmarkEachListItemWithAlloc/FooList-16              187611        187790        +0.10%
BenchmarkEachListItemWithAlloc/SampleList-16           191006        194047        +1.59%
BenchmarkEachListItemWithAlloc/SampleList-16           193163        193787        +0.32%
BenchmarkEachListItemWithAlloc/RawExtensionList-16     108828        107098        -1.59%
BenchmarkEachListItemWithAlloc/RawExtensionList-16     108374        106958        -1.31%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     2888          101885        +3427.87%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     2995          103057        +3340.97%

benchmark                                              old allocs     new allocs     delta
BenchmarkEachListItem/FooList-16                       1              1              +0.00%
BenchmarkEachListItem/FooList-16                       1              1              +0.00%
BenchmarkEachListItem/SampleList-16                    1              1              +0.00%
BenchmarkEachListItem/SampleList-16                    1              1              +0.00%
BenchmarkEachListItem/RawExtensionList-16              1              1              +0.00%
BenchmarkEachListItem/RawExtensionList-16              1              1              +0.00%
BenchmarkEachListItem/UnstructuredList-16              0              0              +0.00%
BenchmarkEachListItem/UnstructuredList-16              0              0              +0.00%
BenchmarkEachListItemWithAlloc/FooList-16              1001           1001           +0.00%
BenchmarkEachListItemWithAlloc/FooList-16              1001           1001           +0.00%
BenchmarkEachListItemWithAlloc/SampleList-16           1001           1001           +0.00%
BenchmarkEachListItemWithAlloc/SampleList-16           1001           1001           +0.00%
BenchmarkEachListItemWithAlloc/RawExtensionList-16     1001           1001           +0.00%
BenchmarkEachListItemWithAlloc/RawExtensionList-16     1001           1001           +0.00%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     0              1001           +Inf%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     0              1001           +Inf%

benchmark                                              old bytes     new bytes     delta
BenchmarkEachListItem/FooList-16                       8             8             +0.00%
BenchmarkEachListItem/FooList-16                       8             8             +0.00%
BenchmarkEachListItem/SampleList-16                    8             8             +0.00%
BenchmarkEachListItem/SampleList-16                    8             8             +0.00%
BenchmarkEachListItem/RawExtensionList-16              8             8             +0.00%
BenchmarkEachListItem/RawExtensionList-16              8             8             +0.00%
BenchmarkEachListItem/UnstructuredList-16              0             0             +0.00%
BenchmarkEachListItem/UnstructuredList-16              0             0             +0.00%
BenchmarkEachListItemWithAlloc/FooList-16              288011        288010        -0.00%
BenchmarkEachListItemWithAlloc/FooList-16              288010        288010        +0.00%
BenchmarkEachListItemWithAlloc/SampleList-16           288010        288010        +0.00%
BenchmarkEachListItemWithAlloc/SampleList-16           288010        288010        +0.00%
BenchmarkEachListItemWithAlloc/RawExtensionList-16     48008         48008         +0.00%
BenchmarkEachListItemWithAlloc/RawExtensionList-16     48008         48008         +0.00%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     0             8016          +Inf%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     0             8016          +Inf%
```

Reply:
> hmm... in the case of Unstructured#EachListItem, we're actually already given a standalone copy of the item, so this change is increasing cost. How much more does using reflection cost us here?

When Unstructured#EachListItem is called, allocNew is false, it will not enter the logic of reflect, and it will only increase a little cost (judging whether allocNew is true)

A standalone Copy will only be generated if the user calls Unstructured#EachListItemWithAlloc

Reply:
`BenchmarkEachListItemWithAlloc/UnstructuredList-16     2995          103057        +3340.97%`

The reason for the deviation is that even if we used UnstructuredList#ListItemWithAlloc, we would not generate a standalone Copy for the user. In fact, this is where this PR was neglected before.

Reply:
that seems really expensive for a shallow copy, doesn't it?

what does it cost if we instead implement UnstructuredList#EachListItemWithAlloc as 
```go
func (u *UnstructuredList) EachListItemWithAlloc(fn func(runtime.Object) error) error {
	for i := range u.Items {
		if err := fn(&Unstructured{Object: u.Items[i].Object}); err != nil {
			return err
		}
	}
	return nil
}
```

Reply:
latest is generated from commit（Add EachListItemWithAlloc variant for unstructuredList）:
https://github.com/sxllwx/kubernetes/commit/1166bb12da36aa83d50d8f7c0e5fc454b5328e18

```bash
benchcmp /tmp/meta_helper.new /tmp/meta_helper.latest
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                                              old ns/op     new ns/op     delta
...
BenchmarkEachListItem/UnstructuredList-16              2619          2853          +8.93%
BenchmarkEachListItem/UnstructuredList-16              2468          2963          +20.06%
...
BenchmarkEachListItemWithAlloc/UnstructuredList-16     101885        31366         -69.21%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     103057        31227         -69.70%

benchmark                                              old allocs     new allocs     delta
...
BenchmarkEachListItem/UnstructuredList-16              0              0              +0.00%
BenchmarkEachListItem/UnstructuredList-16              0              0              +0.00%
...
BenchmarkEachListItemWithAlloc/UnstructuredList-16     1001           1000           -0.10%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     1001           1000           -0.10%

benchmark                                              old bytes     new bytes     delta
...
BenchmarkEachListItem/UnstructuredList-16              0             0             +0.00%
BenchmarkEachListItem/UnstructuredList-16              0             0             +0.00%
...
BenchmarkEachListItemWithAlloc/UnstructuredList-16     8016          8000          -0.20%
BenchmarkEachListItemWithAlloc/UnstructuredList-16     8016          8000          -0.20%
```

> After comparison, it is found that the memory optimization is very small, but the CPU can be improved by 70% (the overhead of using reflect.)